src/libs/ck-libs/ampi/ampi_globals_pie.C

```diff
@@ -0,0 +1,672 @@
+                  from);
+#endif
+      *scan = out;
+    }
+#else
+    // O(log n) method
+    auto iter = my_heap_map.lower_bound((uintptr_t)data);
+    if (iter != my_heap_map.end() &&
+        /* (char *)iter->first <= data && */
+        data <= (char *)iter->first + iter->second.size)
+    {
+      void * out = (char *)iter->second.addr + ((char *)data - (char *)iter->first);
+#if PIEGLOBALS_DEBUG >= 2
+      if (!quietModeRequested)
+        CmiPrintf("pieglobals> [%d][%d][%zu] "
+                  "0x%012" PRIxPTR ": found 0x%012" PRIxPTR ", writing 0x%012" PRIxPTR
+                  " (in %s, to heap)\n",
+                  CmiMyNode(), CmiMyPe(), myrank,
+                  (uintptr_t)scan, (uintptr_t)data, (uintptr_t)out,
+                  from);
+#endif
+      *scan = out;
+    }
+#endif
+  }
+}
+
+static constexpr ptrdiff_t null_funcptr_diff = -1;
+
+template <typename T>
+static void calc_funcptr_diff(
+  ptrdiff_t & diff, T funcptr,
+  const void * objsource, const char * objend,
+  const void * emptyheapstart, const char * allocation)
+{
+  auto ptr = (const char *)funcptr;
+  if (ptr == nullptr)
+  {
+    diff = null_funcptr_diff;
+    return;
+  }
+
+  if (ptr < objsource || objend <= ptr)
+    return;
+
+  diff = ptr - (const char *)objsource + (allocation - (const char *)emptyheapstart);
+}
+
+void AMPI_Rank_Setup(int myrank, int numranks, CmiIsomallocContext ctx)
+{
+  static std::atomic<size_t> rank_count{};
+  const size_t localrank = rank_count++;
+
+#if PIEGLOBALS_DEBUG >= 1
+  if (!quietModeRequested)
+    CmiPrintf("pieglobals> [%d][%d][%zu] AMPI_Rank_Setup(%d, %d, %p)\n",
+              CmiMyNode(), CmiMyPe(), localrank, myrank, numranks, ctx.opaque);
+#endif
+
+  if (ctx.opaque == nullptr)
+    CkAbort("pieglobals requires Isomalloc!");
+
+  if (localrank == 0)
+    pieglobalsdata.nodeleaderrank = myrank;
+
+  // allocate privatized segments for this rank
+
+  std::vector<char *> segment_allocations;
+  for (const auto & obj : pieglobalsdata.objects)
+  {
+#if PIEGLOBALS_DEBUG >= 1
+    if (!quietModeRequested)
+      CmiPrintf("pieglobals> [%d][%d][%d] Processing %s\n",
+                CmiMyNode(), CmiMyPe(), myrank, obj.name.c_str());
+#endif
+
+    auto allocation = (char *)CmiIsomallocContextPermanentAllocAlign(ctx, obj.align, obj.size);
+    segment_allocations.emplace_back(allocation);
+
+    for (const auto & item : obj.items)
+    {
+      char * addr = allocation + item.offset;
+      char * src = obj.source + item.offset;
+      size_t len = item.size;
+
+#if PIEGLOBALS_DEBUG >= 1
+      if (!quietModeRequested)
+        CmiPrintf("pieglobals> [%d][%d][%d] [0x%012" PRIxPTR ", 0x%012" PRIxPTR ") --> [0x%012" PRIxPTR ", 0x%012" PRIxPTR ")\n",
+                  CmiMyNode(), CmiMyPe(), myrank, (uintptr_t)src, (uintptr_t)src + len, (uintptr_t)addr, (uintptr_t)addr + len);
+#endif
+
+      memcpy(addr, src, len);
+    }
+  }
+
+  const CmiIsomallocRegion heapBefore = CmiIsomallocContextGetUsedExtent(ctx);
```

Comment:
It would be nice to have a comment here or maybe even at the beginning of this file explaining the layout of different segments on LInux x86_64.

Reply:
I added a comment that setting PIEGLOBALS_DEBUG to 1 will print this information.